MDEV-37197 - Install PAM modules and systemd units from /lib to /usr/lib

ottok opened the pull request:
## Description

Since Fedora 17 and about 2012, the directory /lib has been a symlink to /usr/lib. Debian started a similar migration in 2019. Nowadays all major Linux distributions have a merged /usr, and the canonical location for files should be /usr/lib instead of just /lib.

The location of PAM modules and systemd files is the last remaining part in MariaDB to fully complete the usr merge migration. Stable releases should not move files around, so target MariaDB 11.8 as the first release to have the /usr merge fully completed.

References:
- https://fedoraproject.org/wiki/Features/UsrMove
- https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#a-merged-usr-is-now-required
- https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/

## Release Notes

* Install PAM modules and systemd units from are now installed to /usr/lib instead of /lib

## How can this PR be tested?

CI should continue to be green for builds, installs and upgrades.

## Basing the PR against the correct MariaDB version

- [x] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

## PR quality check

- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

ottok committed:
MDEV-37197 Deb: Install PAM modules and systemd units in /usr/lib (not /lib)

Since Fedora 17 and about 2012, the directory /lib has been a symlink to
/usr/lib. Debian started a similar migration in 2019. Nowadays all major
Linux distributions have a merged /usr, and the canonical location for
files should be /usr/lib instead of just /lib.

The location of PAM modules and systemd files is the last remaining
`*_DEB` variables should fully complete the usr merge migration.

Additionally re-applied `debputy reformat --style=black` to get files
sorted in a standardized order.

References:
- https://fedoraproject.org/wiki/Features/UsrMove
- https://www.debian.org/releases/bookworm/amd64/release-notes/ch-information.en.html#a-merged-usr-is-now-required
- https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/

ottok commented:
Job [buildbot/amd64-debian-11-debug-ps-embedded](https://buildbot.mariadb.org/#builders/536/builds/27131) failed on test unrelated to changes in this PR:
```
main.vector2                             w8 [ fail ]
        Test ended at 2025-05-25 08:30:34
CURRENT_TEST: main.vector2
Warning: /home/buildbot/amd64-debian-11-debug-ps-embedded/build/libmysqld/examples/mariadb-test-embedded: unknown option '--loose-disable-ssl-verify-server-cert'
Warning: /home/buildbot/amd64-debian-11-debug-ps-embedded/build/libmysqld/examples/mariadb-test-embedded: unknown variable 'loose-ssl-ca=/home/buildbot/amd64-debian-11-debug-ps-embedded/build/mysql-test/std_data/cacert.pem'
Warning: /home/buildbot/amd64-debian-11-debug-ps-embedded/build/libmysqld/examples/mariadb-test-embedded: unknown variable 'loose-ssl-cert=/home/buildbot/amd64-debian-11-debug-ps-embedded/build/mysql-test/std_data/client-cert.pem'
Warning: /home/buildbot/amd64-debian-11-debug-ps-embedded/build/libmysqld/examples/mariadb-test-embedded: unknown variable 'loose-ssl-key=/home/buildbot/amd64-debian-11-debug-ps-embedded/build/mysql-test/std_data/client-key.pem'
Warning: /home/buildbot/amd64-debian-11-debug-ps-embedded/build/libmysqld/examples/mariadb-test-embedded: unknown option '--loose-skip-ssl'
--- /home/buildbot/amd64-debian-11-debug-ps-embedded/build/mysql-test/main/vector2.result	2025-05-25 08:22:14.000000000 +0000
+++ /home/buildbot/amd64-debian-11-debug-ps-embedded/build/mysql-test/main/vector2.reject	2025-05-25 08:30:34.044000000 +0000
@@ -339,7 +339,6 @@
 15	[33]
 2	[33]
 28	[28]
-29	[29]
 3	[33]
 30	[30]
 31	[31]
@@ -349,6 +348,7 @@
 35	[35]
 36	[36]
 37	[37]
+38	[38]
 4	[33]
 5	[33]
 6	[33]
Result content mismatch
```

grooverdan reviewed (approved):
Looks good.

After validating SLES/OpenSUSE and RHEL this can continue on the RPM side (noting the Fedora change https://fedoraproject.org/wiki/Changes/Unify_bin_and_sbin).

This has almost certainly missed the 11.8 stable GA release.

Given the symlinking that has been in place for a long time and services that reference these non-user files don't do so by full path maybe there isn't a concern in changing the install location?

grooverdan commented:
ack.

Interesting not visible in the BCI which I thought was SLES based:
```
$ podman run -ti  registry.suse.com/bci/bci-base:15.7
Trying to pull registry.suse.com/bci/bci-base:15.7...
Getting image source signatures
Copying blob 974449b21a84 done   | 
Copying config 4652df9563 done   | 
Writing manifest to image destination
73bfceae7edc:/ # ls -lad  /bin /sbin /lib*
drwxr-xr-x. 2 root root 4096 May  7 19:18 /bin
drwxr-xr-x. 5 root root 4096 May  7 19:18 /lib
drwxr-xr-x. 3 root root 4096 May  7 19:18 /lib64
drwxr-xr-x. 2 root root 4096 May  7 19:18 /sbin
73bfceae7edc:/ # ls -lad  /usr/bin /usr/sbin /usr/lib*
drwxr-xr-x.  2 root root 12288 May  7 19:18 /usr/bin
drwxr-xr-x. 25 root root  4096 May  7 19:18 /usr/lib
drwxr-xr-x. 10 root root 12288 May  7 19:18 /usr/lib64
drwxr-xr-x.  2 root root  4096 May  7 19:18 /usr/sbin
73bfceae7edc:/ # ls /bin | wc -l
52
73bfceae7edc:/ # ls usr/bin | wc -l
299
``

ottok commented:
CI job [amd64-ubuntu-2204-debug-ps](https://buildbot.mariadb.org/#/builders/534) triggered by Daniel Black failed on unrelated issue:
```
innodb.innodb-lock-inherit-read_commited w3 [ fail ]
        Test ended at 2025-06-13 06:48:32
CURRENT_TEST: innodb.innodb-lock-inherit-read_commited
mysqltest: At line 110: query 'reap' failed with wrong errno ER_LOCK_DEADLOCK (1213): 'Deadlock found when trying to get lock; try restarting transaction', instead of  (0)...
```

grooverdan commented:
Otto, thanks for the PR. Always good to get through migrations.

You mentioned the Debian /usr merge started a while back. If you want to take to to 11.8 or 11.4 that's ok by me if those correspond to the Debian 11/12 (?) version that MariaDB has where the merge/move occurred. There's sufficient tests around PAM and systemd checks in CI to catch if anything is amiss.

As its probably release note worthy the [MDEV-37197](https://jira.mariadb.org/browse/MDEV-37197) has been created for this task.

Can you reword the commit message a bit, you're doing Debian/Ubuntu only changes but reading the commit message it reads looks like you're doing it for everyone. The RPM changes become very dependant on distro which needs more investigation to be done relliably.

grooverdan commented:
> This is modifying `cmake/install_layout.cmake`, hence this is **not** Debian-only. I rebased on latest `main` now.

`INSTALL_SYSTEMD_UNITDIR_DEB` and `INSTALL_PAMDIR_DEB` definitely are Debian only, or applies to those with `-DDEB` or `-DINSTALL_LAYOUT=DEB`.

>  I propose we put this in the dev version first, and backport to 11.4+ in 6-9 months if there are no regressions. I would prefer not to put this change directly in a maintenance branch as it is not a critical bugfix.

sure, ok.

ottok commented:
Ack, commit prefixed with `MDEV-37197` and body explains it is only for Debian.

ottok commented:
Rebased on latest 'main' now.

CI job [continuous-integration/appveyor/pr](https://ci.appveyor.com/project/rasmushoj/server/builds/53086951) failed on:

```
main.sp-bugs2                            w7 [ fail ]  timeout after 240 seconds
        Test ended at 2025-11-17 10:20:26
Test case timeout after 240 seconds
```

CI job [buildbot/amd64-ubuntu-2204-debug-ps](https://buildbot.mariadb.org/#/builders/534/builds/31977/steps/6/logs/stdio) failed on:

```
innodb.gap_locks                         w17 [ fail ]
        Test ended at 2025-11-17 08:03:34
CURRENT_TEST: innodb.gap_locks
--- /home/buildbot/amd64-ubuntu-2204-debug-ps/build/mysql-test/suite/innodb/r/gap_locks.result	2025-11-17 07:49:39.000000000 +0000
+++ /home/buildbot/amd64-ubuntu-2204-debug-ps/build/mysql-test/suite/innodb/r/gap_locks.reject	2025-11-17 08:03:34.403062648 +0000
@@ -8,28 +8,28 @@
 DELETE FROM t1 WHERE b='2' AND c=2;
 SHOW ENGINE INNODB STATUS;
 Type	Name	Status
-InnoDB		2 lock struct(s), 1 row lock(s)
+InnoDB		0 lock struct(s), 0 row lock(s)
 ROLLBACK;
 SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
 BEGIN;
 DELETE FROM t1 WHERE b='2' AND c=2;
 SHOW ENGINE INNODB STATUS;
 Type	Name	Status
-InnoDB		2 lock struct(s), 1 row lock(s)
+InnoDB		0 lock struct(s), 0 row lock(s)
 ROLLBACK;
 SET TRANSACTION ISOLATION LEVEL READ COMMITTED;
 BEGIN;
 DELETE FROM t1 WHERE b='2' AND c=2;
 SHOW ENGINE INNODB STATUS;
 Type	Name	Status
-InnoDB		1 lock struct(s), 0 row lock(s)
+InnoDB		0 lock struct(s), 0 row lock(s)
 ROLLBACK;
 SET TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;
 BEGIN;
 DELETE FROM t1 WHERE b='2' AND c=2;
 SHOW ENGINE INNODB STATUS;
 Type	Name	Status
-InnoDB		1 lock struct(s), 0 row lock(s)
+InnoDB		0 lock struct(s), 0 row lock(s)
 ROLLBACK;
 SET GLOBAL INNODB_STATUS_OUTPUT_LOCKS = @save_locks;
 DROP TABLE t1;
Result content mismatch
```

Neither are due to my change - both are due to the current 'main' branch already being broken and/or CI having unstable tests.

ottok commented:
Rebased on latest 'main'. Can somebody please approve this? I don't see any objections or reasons not to merge.

ottok commented:
Rebased on latest main.

CI job [buildbot/amd64-msan-clang-20](https://buildbot.mariadb.org/#builders/867/builds/7585) failed on unrelated issue in `main.tmp_space_usage`.

gkodinov reviewed (approved):
This is a preliminary review. Thank you for your work on this!

LGTM. Please stand by for the final review.

grooverdan committed:
MDEV-37197 RPM: Install PAM modules and systemd units in /usr/lib (not /lib)

Since Fedora 17 and about 2012, the directory
/lib{,64} has been a symlink to /usr/lib{,64}.

This has been verified on RHEL8, 9, 10, Fedora
, OpenEuler.

SLES1507, OpenSUSE 1506 and earlier (there is no
OpenSUSE 1507) had the old location for pam but
since their 1600 version they are now standardised.

References:
- https://fedoraproject.org/wiki/Features/UsrMove
- https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MariaDB/server?pullRequest=4065) <br/>All committers have signed the CLA.

grooverdan reviewed (approved):
Since I took so long, introduced conflicts, and really wanted RPM changes to happen for consistency I rebased this and appended the RPM changes.

As 12.3 isn't GA this is the targetted version.
